platform/android/src/style/sources/geojson_source.hpp

```diff
@@ -10,6 +10,43 @@
 namespace mbgl {
 namespace android {
 
+using Callback = std::function<void (GeoJSON)>;
+
+struct FeatureConverter {
+    void convertJson(std::string, ActorRef<Callback>);
+
+    template <class JNIType>
+    void convertObject(jni::Object<JNIType>, ActorRef<Callback>);
+};
+
+struct Update {
+    std::unique_ptr<Actor<Callback>> callback;
+
+    using Updater = std::function<void (ActorRef<FeatureConverter>)>;
+    Updater updateFn;
+
+    virtual ~Update();
+
+    virtual void makeUpdateFn() {};
+};
+
+struct CollectionUpdate : Update {
+    std::shared_ptr<jni::jobject> object;
```

Comment:
You have the separate `CollectionUpdate` and `StringUpdate` classes here just to handle the difference in holding on to a different type of object? I think you can probably just get rid of this by capturing the object by value in the lambda (e.g. concentrate all the templating logic on just making different kinds of update function that all have the same signature and capture everything they need to run on their own, and then the `Update` class becomes nothing more than a simple holder for the `updateFn` and the `callback`). I haven't thought this through entirely, but I think you can still get to significantly less code. The templating changes so far definitely look like moving in the right direction, though!

Reply:
Great point. I was able to clean up the templates mess and I'm trying to differentiate the `Update`s only by the lambda that schedules the conversion.